### PR TITLE
Adds schema support and Git capabilities

### PR DESCRIPTION
**Pull request by noelbundick-msft**
Adds schema support
* VS Code extension for YAML validation
* Generate TypeScript interfaces from schema
* Update schemas to validate in strict mode
* Add ability to load and parse YAML files against schema
* Tests now validate all test YAML definitions

Adds Git capabilities
* Add ability to sparse checkout (clone only specific files versus the entire repo) repositories
* Add to Dockerfile

Add sample entities
* ApplicationAssignment
* Cluster

### CODE REVIEW

File: test/control-plane/clusters/cluster1.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+kind: Cluster
+metadata:
+  name: cluster-1
```
Comment (joshuaphelpsms) [inline test/control-plane/clusters/cluster1.yaml:3]: The name of the cluster is different from the one referenced in `test/control-plane/assignments/assignment1.yaml`. cluster1 vs. cluster-1. Should those be consistent?

File: test/schema.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import controlPlaneSchema = require('../schemas/controlplane.json');
+import templateSchema = require('../schemas/template.json');
+
+describe('schemas', () => {
+  it('has well-formed application test data', async () => {
+    await loadAll<Application>('test/applications', applicationSchema);
+  });
+
+  it('has well-formed control plane test data', async () => {
+    await loadAll<Application>(
```
Comment (joshuaphelpsms) [inline test/schema.test.ts:17]: I'm a little confused on why the return type of `Application` is used in all of these tests. Should the return type be ApplicationDeployment on this on and so on? Or am I missing something?

**Review by joshuaphelpsms — APPROVED** [review]
lgtm

File: test/schema.test.ts (hunk 2 above)
Comment (noelbundick-msft) [reply test/schema.test.ts]: Thanks for catching this, fixed!

Added a note for later to see if we can make this a compile-time error. The actual schema check works, but I don't like that this passes with a mismatched generic type vs schema

**Pull request merged**